File: solutions/typescript/006-reverse-integer.ts
/**
 * Problem 6: Reverse Integer (LeetCode 7)
 * Difficulty: Med
 * Language: TypeScript
 */
function reverse(x: number): number {
  const sign = x < 0 ? -1 : 1;
  const reversed = Math.abs(x)
    .toString()
    .split('')
    .reduce((acc, d) => acc + d, '')
    .split('')
    .reverse()
    .join('');
  const result = sign * parseInt(reversed, 10);
  return result > 2**31 - 1 || result < -(2**31) ? 0 : result;
}

File: solutions/scala/006-reverse-integer.scala
/**
 * Problem 6: Reverse Integer (LeetCode 7)
 * Difficulty: Med
 * Language: Scala
 */
def reverse(x: Int): Int = {
  val sign = if (x < 0) -1L else 1L
  val reversed = math.abs(x.toLong).toString.reverse.toLong * sign
  if (reversed > Int.MaxValue || reversed < Int.MinValue) 0
  else reversed.toInt
}
